File: frontend/src/game/rules.ts
import { BROWN_NUMBERS } from "../boardData";
import {
  ACTING_COLORS,
  ALL_DIE_COLORS,
  effectiveValue,
  type ActingColor,
  type DieColor,
  type DieRuntime,
  type GameState,
  type PlayerBoard,
} from "./types";

/** A random integer die value between 1 and 6. */
export function rollValue(): number {
  return Math.floor(Math.random() * 6) + 1;
}

export interface LegalResult {
  /** Highlighted destination cell ids (unprefixed). */
  legal: string[];
  /** Maximum number of cells the player may pick (turquoise cap; 1 otherwise). */
  maxPick: number;
}

/** Context describing how a move is being evaluated. */
export interface MoveContext {
  mode: "active" | "passive";
  board: PlayerBoard;
  dice: Record<DieColor, DieRuntime>;
  /** Active yellow uses this as the row. Ignored in passive mode. */
  round: number;
  /** The die being played; used to exclude it from the passive turquoise count. */
  selectedColor: DieColor;
}

const TURQUOISE_ROWS = [1, 2, 3, 4, 5];
const BLUE_RIGHT = [8, 9, 10, 11, 12, 13];
const BLUE_LEFT = [6, 5, 4, 3, 2, 1];

/** Passive-yellow destination per die value: manche is irrelevant. */
const PASSIVE_YELLOW_CELL: Record<number, string> = {
  1: "yellow-r3-c1",
  2: "yellow-r3-c2",
  3: "yellow-r2-c3",
  4: "yellow-r2-c4",
  5: "yellow-r1-c5",
  6: "yellow-r1-c6",
};

/** The six cells reachable by a passive yellow die (for styling on both boards). */
export const PASSIVE_YELLOW_CELLS: string[] = Object.values(PASSIVE_YELLOW_CELL);

/** The blue sum is always darkblue + white (effective values, so jokers count). */
export function blueSum(dice: Record<DieColor, DieRuntime>): number {
  return effectiveValue(dice.darkblue) + effectiveValue(dice.white);
}

interface BranchInfo {
  nextFree: string | null;
  ref: number;
}

function branchInfo(
  values: Record<string, number>,
  positions: number[]
): BranchInfo {
  let ref = 7;
  let nextFree: string | null = null;
  for (const p of positions) {
    const id = `blue-cell-${p}`;
    if (values[id] !== undefined) {
      ref = values[id];
    } else {
      nextFree = id;
      break;
    }
  }
  return { nextFree, ref };
}

function legalBlue(ctx: MoveContext): string[] {
  const sum = blueSum(ctx.dice);
  const values = ctx.board.values;
  const legal: string[] = [];
  const right = branchInfo(values, BLUE_RIGHT);
  if (right.nextFree && (sum === right.ref + 1 || sum === 7)) {
    legal.push(right.nextFree);
  }
  const left = branchInfo(values, BLUE_LEFT);
  if (left.nextFree && (sum === left.ref - 1 || sum === 7)) {
    legal.push(left.nextFree);
  }
  return legal;
}

/** First empty pink cell, left to right, or null when the track is full. */
export function firstEmptyPink(values: Record<string, number>): string | null {
  for (let n = 1; n <= 12; n++) {
    const id = `pink-cell-${n}`;
    if (values[id] === undefined) return id;
  }
  return null;
}

/**
 * Legal destinations for a die acting as a given color with a given value, in a
 * given context (active or passive). Pure: never mutates state.
 */
export function legalDestinations(
  actingColor: ActingColor,
  value: number,
  ctx: MoveContext
): LegalResult {
  const { board } = ctx;
  switch (actingColor) {
    case "yellow": {
      const id =
        ctx.mode === "passive"
          ? PASSIVE_YELLOW_CELL[value]
          : `yellow-r${ctx.round}-c${value}`;
      return { legal: id && !board.checks[id] ? [id] : [], maxPick: 1 };
    }
    case "turquoise": {
      const free = TURQUOISE_ROWS.map((r) => `turquoise-r${r}-c${value}`).filter(
        (id) => !board.checks[id]
      );
      let sameValue: number;
      if (ctx.mode === "passive") {
        sameValue = ALL_DIE_COLORS.filter(
          (c) =>
            c !== ctx.selectedColor &&
            ctx.dice[c].location === "discarded" &&
            effectiveValue(ctx.dice[c]) === value
        ).length;
      } else {
        sameValue = board.chosenThisTurn.filter((d) => d.value === value).length;
      }
      const cap = 1 + sameValue;
      const maxPick = Math.min(cap, free.length);
      return { legal: free, maxPick };
    }
    case "pink": {
      const target = firstEmptyPink(board.values);
      return { legal: target ? [target] : [], maxPick: 1 };
    }
    case "darkblue": {
      return { legal: legalBlue(ctx), maxPick: 1 };
    }
    case "brown": {
      const start = board.brownLastChecked ?? 0;
      const legal: string[] = [];
      for (let i = 0; i < BROWN_NUMBERS.length; i++) {
        const n = i + 1;
        if (n <= start) continue;
        const id = `brown-cell-${n}`;
        if (BROWN_NUMBERS[i] === value && !board.checks[id]) {
          legal.push(id);
        }
      }
      return { legal, maxPick: 1 };
    }
    default:
      return { legal: [], maxPick: 1 };
  }
}

/** The value written for a pink move when taking the bonus (half, rounded up). */
export function pinkValue(dieValue: number): number {
  return Math.ceil(dieValue / 2);
}

/** The value written for a pink move when taking the points (value x multiplier). */
export function pinkPoints(effectiveValue: number, multiplier: number): number {
  return effectiveValue * multiplier;
}

// ---- Immediate colored bonus-die placement helpers ----

/** Yellow bonus die: any unchecked yellow cell in the die's column (free row). */
export function bonusYellowLegal(board: PlayerBoard, value: number): string[] {
  const legal: string[] = [];
  for (let r = 1; r <= 3; r++) {
    const id = `yellow-r${r}-c${value}`;
    if (!board.checks[id]) legal.push(id);
  }
  return legal;
}

/** Turquoise bonus die: check any one unchecked turquoise cell. */
export function allUncheckedTurquoise(board: PlayerBoard): string[] {
  const legal: string[] = [];
  for (const r of TURQUOISE_ROWS) {
    for (let c = 1; c <= 6; c++) {
      const id = `turquoise-r${r}-c${c}`;
      if (!board.checks[id]) legal.push(id);
    }
  }
  return legal;
}

/** Brown bonus die: a legal brown cell of the chosen number, respecting progression. */
export function bonusBrownLegal(board: PlayerBoard, value: number): string[] {
  const start = board.brownLastChecked ?? 0;
  const legal: string[] = [];
  for (let i = 0; i < BROWN_NUMBERS.length; i++) {
    const n = i + 1;
    if (n <= start) continue;
    const id = `brown-cell-${n}`;
    if (BROWN_NUMBERS[i] === value && !board.checks[id]) legal.push(id);
  }
  return legal;
}

/** A single dark-blue bonus option: which cell to fill and the value to write. */
export interface BlueBonusOption {
  cellId: string;
  value: number;
}

/**
 * Dark-blue bonus die: write a regulation value into the next-free cell of either
 * branch. Each open branch accepts its stepped value (ref-1 left / ref+1 right)
 * and the wildcard 7.
 */
export function blueBonusOptions(board: PlayerBoard): BlueBonusOption[] {
  const values = board.values;
  const options: BlueBonusOption[] = [];
  const push = (cellId: string | null, value: number) => {
    if (!cellId || value < 1) return;
    if (options.some((o) => o.cellId === cellId && o.value === value)) return;
    options.push({ cellId, value });
  };
  const left = branchInfo(values, BLUE_LEFT);
  push(left.nextFree, left.ref - 1);
  push(left.nextFree, 7);
  const right = branchInfo(values, BLUE_RIGHT);
  push(right.nextFree, right.ref + 1);
  push(right.nextFree, 7);
  return options;
}

/** Whether a colored bonus die can still be placed for at least one value. */
export function bonusHasAnyPlaceableValue(
  board: PlayerBoard,
  color: "yellow" | "brown" | "pink"
): boolean {
  if (color === "pink") return firstEmptyPink(board.values) !== null;
  return [1, 2, 3, 4, 5, 6].some((v) =>
    color === "yellow"
      ? bonusYellowLegal(board, v).length > 0
      : bonusBrownLegal(board, v).length > 0
  );
}

/** Availability of each acting color for a die's value in the given context. */
export function colorAvailability(
  value: number,
  ctx: MoveContext
): Record<ActingColor, boolean> {
  const result = {} as Record<ActingColor, boolean>;
  for (const color of ACTING_COLORS) {
    result[color] = legalDestinations(color, value, ctx).legal.length > 0;
  }
  return result;
}

/** Whether a die (by permanent color) has any legal move in the given context. */
export function dieHasAnyLegalMove(color: DieColor, ctx: MoveContext): boolean {
  const value = effectiveValue(ctx.dice[color]);
  const localCtx: MoveContext = { ...ctx, selectedColor: color };
  if (color === "white") {
    return ACTING_COLORS.some(
      (c) => legalDestinations(c, value, localCtx).legal.length > 0
    );
  }
  return legalDestinations(color, value, localCtx).legal.length > 0;
}

/** Build a move context for the active player's board. */
export function activeContext(
  state: GameState,
  player: number,
  round: number,
  selectedColor: DieColor
): MoveContext {
  return {
    mode: "active",
    board: state.boards[player as 1 | 2],
    dice: state.dice,
    round,
    selectedColor,
  };
}

/** Build a move context for the passive player's board. */
export function passiveContext(
  state: GameState,
  player: number,
  selectedColor: DieColor
): MoveContext {
  return {
    mode: "passive",
    board: state.boards[player as 1 | 2],
    dice: state.dice,
    round: 0,
    selectedColor,
  };
}

/** Whether any available die has a legal move for the active player. */
export function anyAvailableDieHasMove(
  state: GameState,
  player: number,
  round: number
): boolean {
  return ALL_DIE_COLORS.some((color) => {
    if (state.dice[color].location !== "available") return false;
    return dieHasAnyLegalMove(color, activeContext(state, player, round, color));
  });
}

/** Whether any discarded (grey-square) die has a legal passive move. */
export function anyDiscardedDieHasMove(state: GameState, player: number): boolean {
  return ALL_DIE_COLORS.some((color) => {
    if (state.dice[color].location !== "discarded") return false;
    return dieHasAnyLegalMove(color, passiveContext(state, player, color));
  });
}

/** Whether any chosen (active-slot) die has a legal passive move. */
export function anyChosenDieHasMove(state: GameState, player: number): boolean {
  return ALL_DIE_COLORS.some((color) => {
    if (state.dice[color].location !== "chosen") return false;
    return dieHasAnyLegalMove(color, passiveContext(state, player, color));
  });
}

/** Whether any of the six dice has a legal passive move. */
export function anyDieHasPassiveMove(state: GameState, player: number): boolean {
  return anyDiscardedDieHasMove(state, player) || anyChosenDieHasMove(state, player);
}

File: backend/game_engine/rules.py
"""Pure legal-destination helpers. Never mutate state. Never roll dice."""

from __future__ import annotations

import math
from typing import Any, Literal, TypedDict

from .types import (
    ACTING_COLORS,
    ALL_DIE_COLORS,
    BROWN_NUMBERS,
    ActingColor,
    DieColor,
    DieRuntime,
    GameState,
    PlayerBoard,
    PlayerId,
    effective_value,
)

TURQUOISE_ROWS = [1, 2, 3, 4, 5]
BLUE_RIGHT = [8, 9, 10, 11, 12, 13]
BLUE_LEFT = [6, 5, 4, 3, 2, 1]

PASSIVE_YELLOW_CELL: dict[int, str] = {
    1: "yellow-r3-c1",
    2: "yellow-r3-c2",
    3: "yellow-r2-c3",
    4: "yellow-r2-c4",
    5: "yellow-r1-c5",
    6: "yellow-r1-c6",
}
PASSIVE_YELLOW_CELLS: list[str] = list(PASSIVE_YELLOW_CELL.values())


class LegalResult(TypedDict):
    legal: list[str]
    max_pick: int


class MoveContext(TypedDict):
    mode: Literal["active", "passive"]
    board: PlayerBoard
    dice: dict[DieColor, DieRuntime]
    round: int
    selected_color: DieColor


class BlueBonusOption(TypedDict):
    cell_id: str
    value: int


def blue_sum(dice: dict[DieColor, DieRuntime]) -> int:
    return effective_value(dice["darkblue"]) + effective_value(dice["white"])


def _branch_info(values: dict[str, int], positions: list[int]) -> tuple[str | None, int]:
    ref = 7
    next_free: str | None = None
    for p in positions:
        cell_id = f"blue-cell-{p}"
        if cell_id in values:
            ref = values[cell_id]
        else:
            next_free = cell_id
            break
    return next_free, ref


def legal_blue(ctx: MoveContext) -> list[str]:
    total = blue_sum(ctx["dice"])
    values = ctx["board"]["values"]
    legal: list[str] = []
    next_free, ref = _branch_info(values, BLUE_RIGHT)
    if next_free and (total == ref + 1 or total == 7):
        legal.append(next_free)
    next_free, ref = _branch_info(values, BLUE_LEFT)
    if next_free and (total == ref - 1 or total == 7):
        legal.append(next_free)
    return legal


def first_empty_pink(values: dict[str, int]) -> str | None:
    for n in range(1, 13):
        cell_id = f"pink-cell-{n}"
        if cell_id not in values:
            return cell_id
    return None


def turquoise_companion_count(ctx: MoveContext, value: int) -> int:
    """Number of dice sharing ``value`` with the selected die in the turquoise group.

    Active games count ``chosen_this_turn``; passive / +1 games count dice sharing the
    selected die's location (grey square or slot). The selected die is never its own
    companion. This is the engine's own turquoise grouping, exposed for the policy.
    """
    if ctx["mode"] == "passive":
        group = ctx["dice"][ctx["selected_color"]]["location"]
        return sum(
            1
            for c in ALL_DIE_COLORS
            if c != ctx["selected_color"]
            and ctx["dice"][c]["location"] == group
            and effective_value(ctx["dice"][c]) == value
        )
    return sum(1 for d in ctx["board"]["chosen_this_turn"] if d["value"] == value)


def legal_destinations(acting_color: ActingColor, value: int, ctx: MoveContext) -> LegalResult:
    board = ctx["board"]
    if acting_color == "yellow":
        cell_id = (
            PASSIVE_YELLOW_CELL.get(value)
            if ctx["mode"] == "passive"
            else f"yellow-r{ctx['round']}-c{value}"
        )
        legal = [cell_id] if cell_id and not board["checks"].get(cell_id) else []
        return {"legal": legal, "max_pick": 1}
    if acting_color == "turquoise":
        free = [
            f"turquoise-r{r}-c{value}"
            for r in TURQUOISE_ROWS
            if not board["checks"].get(f"turquoise-r{r}-c{value}")
        ]
        same_value = turquoise_companion_count(ctx, value)
        cap = 1 + same_value
        return {"legal": free, "max_pick": min(cap, len(free))}
    if acting_color == "pink":
        target = first_empty_pink(board["values"])
        return {"legal": [target] if target else [], "max_pick": 1}
    if acting_color == "darkblue":
        return {"legal": legal_blue(ctx), "max_pick": 1}
    if acting_color == "brown":
        start = board["brown_last_checked"] or 0
        legal: list[str] = []
        for i, printed in enumerate(BROWN_NUMBERS):
            n = i + 1
            if n <= start:
                continue
            cell_id = f"brown-cell-{n}"
            if printed == value and not board["checks"].get(cell_id):
                legal.append(cell_id)
        return {"legal": legal, "max_pick": 1}
    return {"legal": [], "max_pick": 1}


def pink_value(die_value: int) -> int:
    return math.ceil(die_value / 2)


def pink_points(eff: int, multiplier: int) -> int:
    return eff * multiplier


def bonus_yellow_legal(board: PlayerBoard, value: int) -> list[str]:
    return [f"yellow-r{r}-c{value}" for r in range(1, 4) if not board["checks"].get(f"yellow-r{r}-c{value}")]


def all_unchecked_turquoise(board: PlayerBoard) -> list[str]:
    legal: list[str] = []
    for r in TURQUOISE_ROWS:
        for c in range(1, 7):
            cell_id = f"turquoise-r{r}-c{c}"
            if not board["checks"].get(cell_id):
                legal.append(cell_id)
    return legal


def bonus_brown_legal(board: PlayerBoard, value: int) -> list[str]:
    start = board["brown_last_checked"] or 0
    legal: list[str] = []
    for i, printed in enumerate(BROWN_NUMBERS):
        n = i + 1
        if n <= start:
            continue
        cell_id = f"brown-cell-{n}"
        if printed == value and not board["checks"].get(cell_id):
            legal.append(cell_id)
    return legal


def blue_bonus_options(board: PlayerBoard) -> list[BlueBonusOption]:
    values = board["values"]
    options: list[BlueBonusOption] = []

    def push(cell_id: str | None, value: int) -> None:
        if not cell_id or value < 1:
            return
        if any(o["cell_id"] == cell_id and o["value"] == value for o in options):
            return
        options.append({"cell_id": cell_id, "value": value})

    next_free, ref = _branch_info(values, BLUE_LEFT)
    push(next_free, ref - 1)
    push(next_free, 7)
    next_free, ref = _branch_info(values, BLUE_RIGHT)
    push(next_free, ref + 1)
    push(next_free, 7)
    return options


def bonus_has_any_placeable_value(board: PlayerBoard, color: Literal["yellow", "brown", "pink"]) -> bool:
    if color == "pink":
        return first_empty_pink(board["values"]) is not None
    return any(
        len(bonus_yellow_legal(board, v) if color == "yellow" else bonus_brown_legal(board, v)) > 0
        for v in range(1, 7)
    )


def color_availability(value: int, ctx: MoveContext) -> dict[ActingColor, bool]:
    return {c: len(legal_destinations(c, value, ctx)["legal"]) > 0 for c in ACTING_COLORS}


def die_has_any_legal_move(color: DieColor, ctx: MoveContext) -> bool:
    value = effective_value(ctx["dice"][color])
    local: MoveContext = {**ctx, "selected_color": color}
    if color == "white":
        return any(len(legal_destinations(c, value, local)["legal"]) > 0 for c in ACTING_COLORS)
    return len(legal_destinations(color, value, local)["legal"]) > 0  # type: ignore[arg-type]


def active_context(state: GameState, player: int, round_n: int, selected_color: DieColor) -> MoveContext:
    return {
        "mode": "active",
        "board": state["boards"][player],  # type: ignore[index]
        "dice": state["dice"],
        "round": round_n,
        "selected_color": selected_color,
    }


def passive_context(state: GameState, player: int, selected_color: DieColor) -> MoveContext:
    return {
        "mode": "passive",
        "board": state["boards"][player],  # type: ignore[index]
        "dice": state["dice"],
        "round": 0,
        "selected_color": selected_color,
    }


def any_available_die_has_move(state: GameState, player: int, round_n: int) -> bool:
    return any(
        state["dice"][c]["location"] == "available"
        and die_has_any_legal_move(c, active_context(state, player, round_n, c))
        for c in ALL_DIE_COLORS
    )


def any_discarded_die_has_move(state: GameState, player: int) -> bool:
    return any(
        state["dice"][c]["location"] == "discarded"
        and die_has_any_legal_move(c, passive_context(state, player, c))
        for c in ALL_DIE_COLORS
    )


def any_chosen_die_has_move(state: GameState, player: int) -> bool:
    return any(
        state["dice"][c]["location"] == "chosen"
        and die_has_any_legal_move(c, passive_context(state, player, c))
        for c in ALL_DIE_COLORS
    )


def any_die_has_passive_move(state: GameState, player: int) -> bool:
    return any_discarded_die_has_move(state, player) or any_chosen_die_has_move(state, player)
